dali/pipeline/operators/decoder/cache/decoder_cache_blob.cc

```diff
@@ -0,0 +1,129 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operators/decoder/cache/decoder_cache_blob.h"
+#include <fstream>
+#include <mutex>
+#include <unordered_map>
+#include "dali/error_handling.h"
+#include "dali/kernels/alloc.h"
+#include "dali/kernels/span.h"
+#include "dali/pipeline/data/backend.h"
+
+namespace dali {
+
+DecoderCacheBlob::DecoderCacheBlob(std::size_t cache_size,
+                                   std::size_t image_size_threshold,
+                                   bool stats_enabled)
+    : cache_size_(cache_size)
+    , image_size_threshold_(image_size_threshold)
+    , stats_enabled_(stats_enabled) {
+  DALI_ENFORCE(image_size_threshold <= cache_size_, "Cache size should fit at least one image");
+
+  buffer_ = kernels::memory::alloc_unique<uint8_t>(kernels::AllocType::GPU, cache_size_);
+  DALI_ENFORCE(buffer_ != nullptr);
+  tail_ = buffer_.get();
+  buffer_end_ = buffer_.get() + cache_size_;
+  LOG_LINE << "cache size is " << cache_size_ / (1024 * 1024) << " MB" << std::endl;
+}
+
+DecoderCacheBlob::~DecoderCacheBlob() {
+  if (stats_enabled_ && images_seen() > 0) print_stats();
+}
+
+bool DecoderCacheBlob::IsCached(const ImageKey& image_key) const {
+  std::lock_guard<std::mutex> lock(mutex_);
+  return cache_.find(image_key) != cache_.end();
+}
+
+const DecoderCache::ImageShape& DecoderCacheBlob::GetShape(const ImageKey& image_key) const {
+  std::lock_guard<std::mutex> lock(mutex_);
+  const auto it = cache_.find(image_key);
+  DALI_ENFORCE(it != cache_.end(), "cache entry [" + image_key + "] not found");
+  return it->second.dims;
+}
+
+void DecoderCacheBlob::CopyData(const ImageKey& image_key, void* destination_buffer,
+                                cudaStream_t stream) const {
+  std::lock_guard<std::mutex> lock(mutex_);
+  LOG_LINE << "CopyData: image_key[" << image_key << "]" << std::endl;
+  DALI_ENFORCE(!image_key.empty());
+  DALI_ENFORCE(destination_buffer != nullptr);
+  const auto it = cache_.find(image_key);
+  DALI_ENFORCE(it != cache_.end(), "cache entry [" + image_key + "] not found");
+  const auto& data = it->second.data;
+  DALI_ENFORCE(data.data() < tail_);
+  DALI_ENFORCE(data.data() + data.size() <= tail_);
+  MemCopy(destination_buffer, data.data(), data.size(), stream);
+
+  if (stats_enabled_) stats_[image_key].reads++;
+}
+
+void DecoderCacheBlob::Add(const ImageKey& image_key, const uint8_t* data,
+                           const ImageShape& data_shape, cudaStream_t stream) {
+  const std::size_t data_size = volume(data_shape);
+
+  if (stats_enabled_) stats_[image_key].decodes++;
+
+  if (data_size < image_size_threshold_) return;
+
+  if (IsCached(image_key)) return;
+
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Can we move this lock a bit lower in the code, or we need to guard everything here?

Reply:
Also I'm bit afraid of DecoderCacheBlob::Add which takes a lock and also calls IsCached with the lock inside. When someone will rework this code he can easily move IsCached and cause the deadlock.

Reply:
will fix